awx/api/serializers.py

```diff
@@ -2499,8 +2497,7 @@ def to_representation(self, data):
                 if 'help_text' in field:
                     field['help_text'] = _(field['help_text'])
                 if field['type'] == 'become_method':
-                    field.pop('type')
-                    field['choices'] = list(map(operator.itemgetter(0), CHOICES_PRIVILEGE_ESCALATION_METHODS))
+                    field['type'] = 'string'
```

Comment:
So.  `<pulls up chair>`

This logic is here as part of a hack that @wwitzel3 and I are responsible for, and we can probably lose a lot of this complexity now.

When a `CredentialType` is defined, it has a "schema" we store in the database, and that schema has a list of fields it takes:

https://github.com/ansible/awx/blob/devel/awx/main/models/credential/__init__.py#L707

Part of that definition for some fields - like `become_method` - is a list of _choices_.  This means that as time went on, every time Ansible supported a _new_ value to the list of choices, we'd have to write code that would _update_ this schema in the database so that it reflected the _new_ list of possible options (`sudo`, `su`, etc...).

Gross.

So @wwitzel3 and I wrote some code that made a _new_ special type of field, which was `type='become_method'`:

https://github.com/ansible/awx/blob/devel/awx/main/models/credential/__init__.py#L739

A field with _that_ specific special type uses the custom logic you see *here* such that it would populate the choices/options from the `CHOICES_PRIVILEGE_ESCALATION_METHODS` constants defined _in code_.  This meant that if a new `become_method` was added in Ansible, we would just add to `CHOICES_PRIVILEGE_ESCALATION_METHODS` and wouldn't have to modify some JSON-serialized schema defined in the database.

Now that there _are_ no choices, we don't have to do _any_ of this.  We can likely lose any code that has special `field['type'] == 'become_method'` logic (like here and https://github.com/ansible/awx/pull/3093/files#diff-ea6da88b8c0cd3fbf7858ae87933b296L728 and https://github.com/ansible/awx/pull/3093/files#diff-ea6da88b8c0cd3fbf7858ae87933b296L729) and _just_ update the actual schema definition:

https://github.com/ansible/awx/blob/devel/awx/main/models/credential/__init__.py#L739

...to be `type: string` instead of `type: become_method`.

Reply:
@elyezer one thing that we'll really want to test here is that this all works in a 3.4 -> 3.5 upgrade.